runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollOverlay.java

```diff
@@ -22,74 +23,321 @@
  * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
  * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- *
  */
 package net.runelite.client.plugins.cluescrolls;
 
+import java.awt.Color;
 import java.awt.Dimension;
+import java.awt.FontMetrics;
 import java.awt.Graphics2D;
 import java.awt.Point;
-import java.time.Duration;
-import java.time.Instant;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 import javax.inject.Inject;
 import net.runelite.api.Client;
+import net.runelite.api.Item;
 import net.runelite.api.ItemComposition;
+import net.runelite.api.ObjectComposition;
+import net.runelite.api.coords.WorldPoint;
+import net.runelite.api.queries.EquipmentItemQuery;
+import net.runelite.client.plugins.cluescrolls.clues.AnagramClue;
+import net.runelite.client.plugins.cluescrolls.clues.CipherClue;
+import net.runelite.client.plugins.cluescrolls.clues.CrypticClue;
+import net.runelite.client.plugins.cluescrolls.clues.EmoteClue;
+import net.runelite.client.plugins.cluescrolls.clues.FairyRingClue;
+import net.runelite.client.plugins.cluescrolls.clues.MapClue;
 import net.runelite.client.ui.overlay.Overlay;
-import net.runelite.client.ui.overlay.OverlayPosition;
 import net.runelite.client.ui.overlay.components.PanelComponent;
+import net.runelite.client.util.QueryRunner;
 
 public class ClueScrollOverlay extends Overlay
 {
-	private static final Duration WAIT_DURATION = Duration.ofMinutes(4);
+	private static final Color TITLED_CONTENT_COLOR = new Color(190, 190, 190);
 
+	private final ClueScrollPlugin plugin;
 	private final Client client;
 	private final PanelComponent panelComponent = new PanelComponent();
 
-	ClueScroll clue;
-	Instant clueTimeout;
+	@Inject
+	private QueryRunner queryRunner;
 
 	@Inject
-	public ClueScrollOverlay(Client client)
+	public ClueScrollOverlay(ClueScrollPlugin plugin, Client client)
 	{
-		setPosition(OverlayPosition.TOP_LEFT);
+		this.plugin = plugin;
 		this.client = client;
 	}
 
 	@Override
 	public Dimension render(Graphics2D graphics, Point parent)
 	{
+		Object clue = plugin.getClue();
+
 		if (clue == null)
 		{
 			return null;
 		}
 
-		if (clueTimeout == null || Instant.now().compareTo(clueTimeout.plus(WAIT_DURATION)) >= 0)
+		panelComponent.getLines().clear();
+
+		if (clue instanceof AnagramClue)
```

Comment:
At least move handling for each type of clue into its own method to make this easier to read.

Reply:
It looks like you could give the clue classes a common `makeClueHint(PanelComponent)` and make this a lot simpler.